### ISSUE TITLE
vms with v0.9.X need interfaces and network section

### ISSUE THREAD

**Issue by karmab** [kubevirt/kubevirt#1618]
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:
>
 /kind bug
> /kind enhancement

**What happened**:
vms lack networking

**What you expected to happen**:
vms have networking

**How to reproduce it (as minimally and precisely as possible)**:
launch a vm with an old yaml definition

**Anything else we need to know?**:

**Environment**:
- KubeVirt version (use `virtctl version`):
Client Version: version.Info{GitVersion:"v0.9.1", GitCommit:"93e8226fa15325a0eaea2b693de82824a8b98270", GitTreeState:"clean", BuildDate:"2018-10-05T07:01:51Z", GoVersion:"go1.10", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{GitVersion:"v0.9.1", GitCommit:"93e8226fa15325a0eaea2b693de82824a8b98270", GitTreeState:"clean", BuildDate:"2018-10-05T07:01:51Z", GoVersion:"go1.10", Compiler:"gc", Platform:"linux/amd64"}

- Kubernetes version (use `kubectl version`):
Client Version: version.Info{Major:"1", Minor:"12", GitVersion:"v1.12.1", GitCommit:"4ed3216f3ec431b140b1d899130a69fc671678f4", GitTreeState:"clean", BuildDate:"2018-10-05T16:46:06Z", GoVersion:"go1.10.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"11+", GitVersion:"v1.11.0+d4cacc0", GitCommit:"d4cacc0", GitTreeState:"clean", BuildDate:"2018-10-11T07:30:01Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}

- VM or VMI specifications:
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

**Comment by rmohr** [comment]
@karmab I suspect that the webhooks could not successfully register in cluster and that `virt-api` has some issues. Could you check the `kube-system` namespace if you see something strange?

**Comment by karmab** [comment]
does this mean this issue should be seen as random ?

**Comment by davidvossel** [comment]
> do you have some thoughts on how we could reject CRDs until our webhooks are fine?

@rmohr ugh... This is bad. We need to set the failure policy on the webhook. By default k8s ignores if the webhook isn't available, which lets CRD objects get posted even if the webhook isn't online.

Essentially our webhooks are just best effort right now and not guaranteed.

**Comment by slintes** [comment]
>  I think we need to move the CRD registration into virt-api

afaik CRD registration can/should be handled by the OLM (Operator Lifecycle manager) later on. Might be worth to discuss this with the operator team?

**Comment by rmohr** [comment]
> afaik CRD registration can/should be handled by the OLM (Operator Lifecycle manager) later on. Might be worth to discuss this with the operator team?

Yeah I don't think so. I think that t here is a misunderstanding. OLM normally need their own CRDs to manage the installation. I have it on my list to clarify that with them. Also since OLM is in a different repo I simply think that it can't do it properly if it would want to do it.

**Comment by rmohr** [comment]
@rthallisey FYI

**Comment by fabiand** [comment]
There are the two CRDs - Operator one, and the one by the application (KubeVirt's CRDs). But I also learned that the Operator wants to control both of them, as @slintes noted.

Could the API server not reject requests until it knows that the webhooks are present and registered properly?

**Comment by rmohr** [comment]
> There are the two CRDs - Operator one, and the one by the application (KubeVirt's CRDs). But I also learned that the Operator wants to control both of them, as @slintes noted.

That would immensely increase the complexity of updates since the operator is external. I hope that they are not serious about that and that it is just a misunderstanding. The CRDs provide a public API but they are application internal.

> Could the API server not reject requests until it knows that the webhooks are present and registered properly?

Not really, that's a chicken- and egg problem. If the webhooks are not registered, it is not involved in CRDs, so any CR can be created.

But as said, if the operator will be part of this repo, then we can hand all that over to the operator.

**Comment by rthallisey** [comment]
> What I originally thought is, that the issue is that the installation only worked partially, so CRDs were created but the webhook registration failed. I think we need to move the CRD registration into virt-api and let it register the CRDs after the webhooks have successfully registered.
>
This would mean moving the CRDs out of yaml and into virt-api.go (or some other file)?  This would make kubevirt 1) an operator by definition and 2) require cluster-admin permissions.  At that point, the delta between kubevirt and kubevirt-operator would be a control loop watching the CRD, OLM integration templates, and operator tests.

**Comment by rthallisey** [comment]
Following up on my comment, I don't think 2) is a valid option.  I don't think we're going to get a green light to make the app cluster-admin.  With that being said, it leaves us either:
 1) Move deployment code to kubevirt-operator
 2) Move kubevirt-operator into kubevirt and rewrite to remove operator-sdk dependency

Is there a technical reason blocking either of our options?

**Comment by rmohr** [comment]
> 1. Move deployment code to kubevirt-operator
> 2. Move kubevirt-operator into kubevirt and rewrite to remove operator-sdk dependency
> 
> Is there a technical reason blocking either of our options?

Both are technically possible. Option (1) would obviously be a pain to maintain and properly test.

**Comment by rmohr** [comment]
> 2. Move kubevirt-operator into kubevirt and rewrite to remove operator-sdk dependency

This option requires us to do daily releases if necessary (I think that is the right thing).

**Comment by rthallisey** [comment]
>>   2. Move kubevirt-operator into kubevirt and rewrite to remove operator-sdk dependency
>>
>This option requires us to do daily releases if necessary (I think that is the right thing).
>
Why would this require daily releases?

>  Option (1) would obviously be a pain to maintain and test properly.
>
I think we all agree from a kubevirt perspective, that there is pain in maintaining & testing deployment code and application code in different places.  There will be some code duplication, ripping out tested code paths, and another repo to maintain.  But, I think there is just as much pain maintaining deployment and application code in the same repo.  For instance, the release cadence will be tied, requiring new kubevirt releases for deployment bug fixing. Second, using the same repo links kubevirt to a deployment pattern for the foreseeable future.  The two prior options, apbs and ansible playbooks, were easy to replace, but a build-in solution will be more difficult to change.

In my experience, I've found that separating deployment from application code is effective.  I've found it to help scale a project.  Clearly both sides have pain points, but in my opinion, application and deployment should be separate.

What are other opinions? @fabiand @rmohr @davidvossel
